### ISSUE TITLE
Question: EfficientAD Pred Scores are very low.

### ISSUE THREAD

**Issue by rishabh-akridata** [open-edge-platform/anomalib#2333]
Hi,
The pred scores of the efficient AD model are coming very low. Any idea, what is the reason behind this?
```
[0.5111125114597613,
 0.5133660426906085,
 0.5584840866012671,
 0.5097309616622074,
 0.5100746147399043,
 0.5101457543058581,
 0.5078011069118742,
 0.49999999867182765,
 0.5171874406178997,
 0.5087297667708703,
 0.5009551808039379,
 0.5057654561524445,
 0.4999895069959258,
 0.539257350719084,
 0.49940360237013637,
 1.0,
 0.4994316863525398,
 0.4994300963089074,
 0.5165583225774157,
 0.5048455843669841,
 0.5154843004558268,
 0.5397329320390526,
 0.5101223970673893,
 0.5083544491795497,
 0.5032673477437677,
 0.49942594466358836,
 0.5006433367892856,
 0.5098894506945904,
 0.5126176016486443,
 0.5127730119775831,
 0.5302341765149362,
 0.49942822048686353,
 0.5073944904970427,
 0.4994255681267349,
 0.5114618296351392,
 0.5280324174412317,
 0.49941536510695117,
 0.5089816112649264,
 0.5227437273384679,
 0.5137013902640468,
 0.5014994171221683,
 0.5085513064539936,
 0.5100113379541017,
 0.5132433620694963,
 0.5113751708183177,
 0.4994075496982797,
 0.5080081078810714,
 0.5129591863193106,
 0.4994130655979363,
 0.4994456381387047,
 0.507350383665099,
 0.6742883947936092,
 0.4994323178984793,
 0.519790254606196,
 0.5131760290454198,
 0.5125366592855003,
 0.5056670881684252,
 0.5067094590581136,
 0.5135149821639905,
 1.0,
 0.4994077042089933,
 0.49940675500850656,
 0.6721566180050411,
 0.5094002476392792,
 0.5132569147398781,
 0.49941373322589055,
 0.5078058847904434,
 0.49941520218447955,
 0.49940407564220735,
 0.5204894080587092,
 0.5146554660286121,
 0.5027675370188384,
 0.5133325426428347,
 0.5122029143733313,
 0.5094226716153291,
 0.515083500550321,
 0.4994165772855578,
 0.5002911459985357,
 0.5089317321960678,
 0.5190379724898744,
 0.499425803213236,
 0.5095397871944086,
 0.6957075571253086]
```
Thanks.

**Comment by alexriedel1** [comment]
What do you mean by low? The scores are normalized in the range 0,1 with 0 bring normal and 1 being abnormal

**Comment by rishabh-akridata** [comment]
@alexriedel1 The scores are normalized between 0 and 1. But they are not high for the anomalous samples. Mostly all the scores are close to 0.5 only.

**Comment by alexriedel1** [comment]
Assuming you have trained your Model using normal and abnormal Images, maybe the difference between normal and abnormal isnt very large. The training procedure tries to normalize the outputs in a way that 0.5 is the threshold for abnormal. You could try to use a different algorithm than efficientad and see of the problem still exists. If yes, then the Chance is high that your normal and abnormal are Not different enough. Maybe you can show some of your Images?

**Comment by rishabh-akridata** [comment]
@alexriedel1 I am getting the following anomaly scores on the MVTEC bottle class. Is this reasonable? I tried other methods such as Padim and the scores are pretty high such as 0.70, and 0.94 for anomalous images. Is this because the Padim is a distance-based method and efficientAD is a reconstruction-based method?
![Screenshot 2024-10-03 at 2 49 44 PM](https://github.com/user-attachments/assets/d457e5f3-16bd-4d2e-8eab-ced2786d807e)

**Comment by alexriedel1** [comment]
How do you obtain these results? Through validation while Training or through testing After training? Can you Share your Code?

**Comment by watertianyi** [comment]
@alexriedel1 
I used patchcore and EfficientAD to classify the same batch of data sets respectively. The F1Score and AUROC of patchcore are much better than EfficientAD, and the heat map of patchcore can clearly reflect the defective area. EfficientAD can hardly reflect the defective area. It was all trained in the past. 100 epochs. I would like to know if there is room for improvement when changing EfficientAD’s epoch to 1000. I don’t have mask annotation data. Is there any room for improvement, especially the defects and accuracy of heat maps?

**Comment by wawatt** [comment]
Have you solved this problem? I encountered the same problem from version 1.1.3 to version 2.0

**Comment by samet-akcay** [comment]
@wawatt, can you elaborate the problem you are experiencing?
Your dataset, model, your evalution criteria? If you could share your code, we would be able to reproduce?

**Comment by wawatt** [comment]
> [@wawatt](https://github.com/wawatt), can you elaborate the problem you are experiencing? Your dataset, model, your evalution criteria? If you could share your code, we would be able to reproduce?

# train.py
from anomalib.data import MVTec
from anomalib.models import EfficientAd
from anomalib.engine import Engine


if __name__ == "__main__":

    # Initialize the datamodule, model and engine
    datamodule = MVTec(
        category="bottle",
        num_workers=0,
        train_batch_size=1,
    )
    model = EfficientAd()
    engine = Engine()

    # Train the model
    engine.fit(datamodule=datamodule, model=model)



#infer.py
# Import the required modules
from anomalib.data import MVTec,PredictDataset
from anomalib.models import EfficientAd
from anomalib.engine import Engine
import cv2

if __name__ == "__main__":
    dataset = PredictDataset(
        path=r"E:\mvtec_anomaly_detection\bottle\test\good\000.png",
        image_size=(256, 256),
    )
    model = EfficientAd()
    engine = Engine()
    predictions = engine.predict(
        dataset=dataset,
        model=model,
        ckpt_path=r"results\EfficientAd\MVTec\bottle\latest\weights\lightning\model.ckpt",
    )

    for prediction in predictions:
        anomaly_maps = prediction["anomaly_maps"]
        anomaly_map = anomaly_maps[0].permute(1, 2, 0).cpu().numpy()
        print(anomaly_map.min(), anomaly_map.max())   # !!!!! the values close to 0.5, onnx output is same

        cv2.imshow("Anomaly Map", anomaly_map)
        cv2.waitKey(0)

**Comment by YiAnXS** [comment]
I'm having the same issue：
from anomalib.models import Patchcore
from anomalib.models import EfficientAd
from anomalib.engine import Engine
from anomalib.data import Folder
from anomalib.loggers import AnomalibMLFlowLogger, AnomalibTensorBoardLogger, AnomalibWandbLogger

from anomalib.data.utils import (
    TestSplitMode,
    ValSplitMode,
)

import torchvision.transforms.v2 as v2

if __name__ == '__main__':
    data_transform = v2.Compose([
        v2.Resize((256, 256)),
        v2.RandomHorizontalFlip(),
        v2.ColorJitter(brightness=(0.1, 1)),
    ])

    # Create the datamodule
    datamodule = Folder(
        name="mvtec",
        root="./datasets/MVTecAD/bottle",
        normal_dir="./train/good",
        abnormal_dir=["./test/broken_large", "./test/broken_small", "./test/contamination"],
        train_batch_size=1,
        eval_batch_size=4,
        num_workers=2,
        normal_test_dir='./test/good',
        mask_dir=["./ground_truth/broken_large", "./ground_truth/broken_small", "./ground_truth/contamination"],
        normal_split_ratio=0.2,
        extensions=None,
        test_split_mode=TestSplitMode.FROM_DIR,
        test_split_ratio=0.3,
        val_split_mode=ValSplitMode.FROM_TEST,
        val_split_ratio=0.5,
        seed=1,
    )

    # Setup the datamodule
    datamodule.setup()

    #model = Patchcore( pretrained_cfg_overlay=dict(file='./pre_trained/wide_resnet50_2-95faca4d.pth'))
    model = EfficientAd()

    engine = Engine(
    logger=AnomalibTensorBoardLogger(save_dir="logs/tensorboard"),
        max_epochs=20,
        default_root_dir='./results',

    )
    engine.train(datamodule=datamodule, model=model)
